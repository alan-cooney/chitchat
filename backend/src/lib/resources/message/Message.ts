import { IGraphqlApi, MappingTemplate } from "@aws-cdk/aws-appsync";
import { AttributeType, BillingMode, Table } from "@aws-cdk/aws-dynamodb";
import { Construct } from "@aws-cdk/core";
import { join } from "path";

export default class Message {
  constructor(scope: Construct, api: IGraphqlApi) {
    const id = "message";

    // Data
    const table = new Table(scope, `${id}Table`, {
      billingMode: BillingMode.PAY_PER_REQUEST,
      partitionKey: {
        name: "pk", // Will be APP#{{app-id}}#CONV#{{conv-id}}
        type: AttributeType.STRING,
      },
      sortKey: {
        name: "id", // ID should be time sortable (i.e. timestamp + short UUID)
        type: AttributeType.STRING,
      },
      pointInTimeRecovery: true,
    });

    const tableDS = api.addDynamoDbDataSource(`${id}TableDataSource`, table);

    // Mutation -> addMessage
    tableDS.createResolver({
      typeName: "Mutation",
      fieldName: "addMessage",
      requestMappingTemplate: MappingTemplate.fromFile(
        join(__dirname, "./velocity/addMessageRequest.vtl")
      ),
      responseMappingTemplate: MappingTemplate.dynamoDbResultItem(),
    });

    // Converation -> messages
    tableDS.createResolver({
      typeName: "Participant",
      fieldName: "conversation",
      requestMappingTemplate: MappingTemplate.fromFile(
        join(__dirname, "./velocity/conversationMessagesRequest.vtl")
      ),
      responseMappingTemplate: MappingTemplate.dynamoDbResultItem(),
    });
  }
}
