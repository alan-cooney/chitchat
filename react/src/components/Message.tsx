import React from "react";
import { Card, CardProps } from "react-bootstrap";
import { DateTime } from "luxon";
import { ConversationViewQuery, User } from "../types/graphql";

export type ConversationMessage = ConversationViewQuery["conversation"]["messages"]["items"][0];

export default function Message({
  message,
  user,
}: {
  message: ConversationMessage;
  user: Pick<User, "id">;
}) {
  const isAuthor = user.id === message.user.id;

  const bg: CardProps["bg"] = isAuthor ? "primary" : "light";
  const text: CardProps["text"] = isAuthor ? "white" : "dark";

  const created = DateTime.fromISO(message.created);
  const createdFloatClass = isAuthor ? "text-end" : "text-start";

  return (
    <div className={isAuthor ? "d-flex flex-row-reverse" : "d-flex"}>
      <div className="p-2">
        <div
          className="bg-secondary rounded-circle"
          style={{ height: "3em", width: "3em" }}
        >
          <span className="d-table" style={{ width: "100%", height: "100%" }}>
            <span className="d-table-cell align-middle text-center text-white font-weight-bold">
              {message.user.givenName.substr(0, 2).toUpperCase()}
            </span>
          </span>
        </div>
      </div>
      <div>
        <Card bg={bg} border={bg} text={text}>
          <Card.Body>{message.text}</Card.Body>
        </Card>
        <div className={`text-muted light small ${createdFloatClass}`}>
          {created.toLocaleString(DateTime.TIME_SIMPLE)}
        </div>
      </div>
    </div>
  );
}
