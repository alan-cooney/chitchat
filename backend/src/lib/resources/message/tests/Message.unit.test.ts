import "@aws-cdk/assert/jest";
import { SynthUtils } from "@aws-cdk/assert";
import { Stack } from "@aws-cdk/core";
import { GraphqlApi } from "@aws-cdk/aws-appsync";
import { AttributeType, Table } from "@aws-cdk/aws-dynamodb";
import Message from "../Message";

/**
 * Test stack
 */
const stack = new Stack();
const api = new GraphqlApi(stack, "api", {
  name: "test",
});
const participantTable = new Table(stack, "participantTable", {
  partitionKey: { name: "pk", type: AttributeType.STRING },
});
new Message(stack, { api, participantTable });

it("creates a dynamodb table", () => {
  expect(stack).toHaveResource("AWS::DynamoDB::Table", {
    BillingMode: "PAY_PER_REQUEST",
    KeySchema: [
      {
        AttributeName: "pk",
        KeyType: "HASH",
      },
      {
        AttributeName: "id",
        KeyType: "RANGE",
      },
    ],
    PointInTimeRecoverySpecification: {
      PointInTimeRecoveryEnabled: true,
    },
  });
});

it("creates resources matching the snapshot", () => {
  const cloudformation = SynthUtils.toCloudFormation(stack);
  expect(cloudformation).toMatchSnapshot();
});
