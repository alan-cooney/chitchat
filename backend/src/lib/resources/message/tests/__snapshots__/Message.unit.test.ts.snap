// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`creates resources matching the snapshot 1`] = `
Object {
  "Resources": Object {
    "apiC8550315": Object {
      "Properties": Object {
        "AuthenticationType": "API_KEY",
        "Name": "test",
      },
      "Type": "AWS::AppSync::GraphQLApi",
    },
    "apiDefaultApiKey6AB8D7C4": Object {
      "DependsOn": Array [
        "apiSchema0EA92056",
      ],
      "Properties": Object {
        "ApiId": Object {
          "Fn::GetAtt": Array [
            "apiC8550315",
            "ApiId",
          ],
        },
      },
      "Type": "AWS::AppSync::ApiKey",
    },
    "apiSchema0EA92056": Object {
      "Properties": Object {
        "ApiId": Object {
          "Fn::GetAtt": Array [
            "apiC8550315",
            "ApiId",
          ],
        },
        "Definition": "",
      },
      "Type": "AWS::AppSync::GraphQLSchema",
    },
    "apiconversationTableDataSourceF25AD297": Object {
      "Properties": Object {
        "ApiId": Object {
          "Fn::GetAtt": Array [
            "apiC8550315",
            "ApiId",
          ],
        },
        "DynamoDBConfig": Object {
          "AwsRegion": Object {
            "Ref": "AWS::Region",
          },
          "TableName": Object {
            "Ref": "conversationTable542DF655",
          },
        },
        "Name": "conversationTableDataSource",
        "ServiceRoleArn": Object {
          "Fn::GetAtt": Array [
            "apiconversationTableDataSourceServiceRole69D326C3",
            "Arn",
          ],
        },
        "Type": "AMAZON_DYNAMODB",
      },
      "Type": "AWS::AppSync::DataSource",
    },
    "apiconversationTableDataSourceMutationaddMessageResolver6F7CDE24": Object {
      "DependsOn": Array [
        "apiconversationTableDataSourceF25AD297",
        "apiSchema0EA92056",
      ],
      "Properties": Object {
        "ApiId": Object {
          "Fn::GetAtt": Array [
            "apiC8550315",
            "ApiId",
          ],
        },
        "DataSourceName": "conversationTableDataSource",
        "FieldName": "addMessage",
        "Kind": "UNIT",
        "RequestMappingTemplate": "#set($input = $context.args.input)
#set($uuid = $util.autoId())
#set($shortUUID = $uuid.substring(8))
#set($id = \\"$util.time.nowEpochSeconds()$shortUUID\\")

{
    \\"version\\": \\"2017-02-28\\",
    \\"operation\\": \\"PutItem\\",
    \\"key\\" : {
        \\"pk\\": $util.dynamodb.toDynamoDBJson(\\"APP#$ctx.args.input.appID#CONV#$ctx.args.input.conversationID\\"),
        \\"id\\": $util.dynamodb.toDynamoDBJson($id)
    },
    \\"attributeValues\\": $util.dynamodb.toMapValuesJson($input)
}",
        "ResponseMappingTemplate": "$util.toJson($ctx.result)",
        "TypeName": "Mutation",
      },
      "Type": "AWS::AppSync::Resolver",
    },
    "apiconversationTableDataSourceParticipantconversationResolver4E52AED7": Object {
      "DependsOn": Array [
        "apiconversationTableDataSourceF25AD297",
        "apiSchema0EA92056",
      ],
      "Properties": Object {
        "ApiId": Object {
          "Fn::GetAtt": Array [
            "apiC8550315",
            "ApiId",
          ],
        },
        "DataSourceName": "conversationTableDataSource",
        "FieldName": "conversation",
        "Kind": "UNIT",
        "RequestMappingTemplate": "{
    \\"version\\" : \\"2017-02-28\\",
    \\"operation\\" : \\"Query\\",
    #if($util.isNullOrEmpty($ctx.args.nextToken) == false)
        \\"nextToken\\" : \\"$ctx.args.nextToken\\",
    #end
    #if($util.isNullOrEmpty($ctx.args.limit) == false)
        \\"limit\\" : $ctx.args.limit,
    #end
    \\"expression\\": \\"#pk = :pk\\",
    \\"expressionAttributeNames\\": {
        \\"#pk\\": \\"pk\\"
    },
    \\"expressionAttributeValues\\": {
        \\":pk\\": \\"APP#$ctx.source.appID#USER#$ctx.source.id\\"
    },
    \\"scanIndexForward\\": false
}",
        "ResponseMappingTemplate": "$util.toJson($ctx.result)",
        "TypeName": "Participant",
      },
      "Type": "AWS::AppSync::Resolver",
    },
    "apiconversationTableDataSourceServiceRole69D326C3": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "appsync.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "apiconversationTableDataSourceServiceRoleDefaultPolicy67B6F07F": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "dynamodb:BatchGetItem",
                "dynamodb:GetRecords",
                "dynamodb:GetShardIterator",
                "dynamodb:Query",
                "dynamodb:GetItem",
                "dynamodb:Scan",
                "dynamodb:ConditionCheckItem",
                "dynamodb:BatchWriteItem",
                "dynamodb:PutItem",
                "dynamodb:UpdateItem",
                "dynamodb:DeleteItem",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "conversationTable542DF655",
                    "Arn",
                  ],
                },
                Object {
                  "Ref": "AWS::NoValue",
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "apiconversationTableDataSourceServiceRoleDefaultPolicy67B6F07F",
        "Roles": Array [
          Object {
            "Ref": "apiconversationTableDataSourceServiceRole69D326C3",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "conversationTable542DF655": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "AttributeDefinitions": Array [
          Object {
            "AttributeName": "pk",
            "AttributeType": "S",
          },
          Object {
            "AttributeName": "id",
            "AttributeType": "S",
          },
        ],
        "BillingMode": "PAY_PER_REQUEST",
        "KeySchema": Array [
          Object {
            "AttributeName": "pk",
            "KeyType": "HASH",
          },
          Object {
            "AttributeName": "id",
            "KeyType": "RANGE",
          },
        ],
        "PointInTimeRecoverySpecification": Object {
          "PointInTimeRecoveryEnabled": true,
        },
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Retain",
    },
  },
}
`;
